fdbclient/NativeAPI.actor.cpp

```diff
@@ -8013,6 +8015,72 @@ ACTOR Future<Version> setPerpetualStorageWiggle(Database cx, bool enable, LockAw
 	return version;
 }
 
+ACTOR Future<Version> checkBlobSubrange(Database db, KeyRange keyRange, Optional<Version> version) {
+	state Transaction tr(db);
+	state Version readVersionOut = invalidVersion;
+	loop {
+		try {
+			wait(success(tr.readBlobGranules(keyRange, 0, version, &readVersionOut)));
+			return readVersionOut;
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+	}
+}
+
+ACTOR Future<Version> verifyBlobRangeActor(Reference<DatabaseContext> cx, KeyRange range, Optional<Version> version) {
+	state Database db(cx);
+	state Transaction tr(db);
+	state Standalone<VectorRef<KeyRangeRef>> allRanges;
+	state Key beginKey = range.begin;
+	state Key endKey = range.end;
+	state Version readVersionOut = invalidVersion;
+	state int batchSize = CLIENT_KNOBS->BG_TOO_MANY_GRANULES / 2;
+	loop {
+		try {
+			wait(store(allRanges, tr.getBlobGranuleRanges(KeyRangeRef(beginKey, endKey), 20 * batchSize)));
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+
+		if (allRanges.empty()) {
+			return readVersionOut;
+		}
+
+		state std::vector<Future<Version>> checkParts;
+		// Chunk up to smaller ranges than this limit. Must be smaller than BG_TOO_MANY_GRANULES to not hit the limit
+		int maxChunkSize = CLIENT_KNOBS->BG_TOO_MANY_GRANULES / 2;
+		KeyRange currentChunk;
+		int currentChunkSize = 0;
+		for (auto& it : allRanges) {
+			if (currentChunkSize == maxChunkSize) {
+				checkParts.push_back(checkBlobSubrange(db, currentChunk, version));
+				currentChunkSize = 0;
+			}
+			if (currentChunkSize == 0) {
+				currentChunk = it;
+			} else if (it.begin != currentChunk.end) {
+				fmt::print("ERROR: Blobrange check failed, gap in blob ranges from [{0} - {1})\n",
```

Comment:
This output was useful for fdbcli, we shouldn't have fmt::prints in the general client code (unless you wanted to set a dbgStdout flag on this call that would be true for fdbcli and false for the api cases or something, that might be the best way to fix it)

Reply:
Also, I think this refactor will also not fail the check if the very start, very end, or boundaries between batches of allRanges of the desired range isn't blobbified.
(eg if i have a range [B - C), and i do a check for [A - C) or [B - D), I think this check will still return successfully instead of returning invalidVersion).